Half of fix for OOM in sync (#13288) (2.1)

dpsutton opened the pull request:
* truncate type/Text fields to 1234 (a distinct number we can use as a
heuristic for longer fields if desired)

* test truncation in metadata-queries-test

* test that fingerprints reflect these truncations in fingerprint-test



###### Before submitting the PR, please make sure you do the following
- [ ] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.
### Tests
-  [ ] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`)
-  [X] If there are changes to the backend codebase, run the backend tests with `lein test && lein lint && ./bin/reflection-linter`

-  [ ] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

dpsutton committed:
Half of fix for OOM in sync (#13288) (2.1)

* truncate type/Text fields to 1234 (a distinct number we can use as a
heuristic for longer fields if desired)

* test truncation in metadata-queries-test

* test that fingerprints reflect these truncations in fingerprint-test

sbelak reviewed (approved):
This looks good. 

One thing to note: this will in large part disable our JSON detection, as all truncated JSON fields will be ill-formed. This can be somewhat mitigated by playing around with `truncation-size`. In addition we could also skip truncated values when detecting JSON.

salsakran commented:
can we look for prefixes that suggest json?
things like "[{" "{", etc?

robdaemon commented:
long term, might be good to treat BLOBs/CLOBs as streams? it would alleviate any OOM concerns there. It would slow sync down a bit

sbelak commented:
@salsakran: good point. The threshold is high enough (95% positive rate), that we can be pretty aggressive with heuristics. I'd be tempted to just use such heuristics and eschew parsing altogether (how we currently detect JSON).

dpsutton commented:
> can you try out if this works with BLOBs and other things we mark as text but are not strings per-se (the Postgres driver has a pretty exhaustive list of things we map to :type/Text)

i'll check those now

Can someone give me some context? What actually uses the JSON percentages? And i'm pretty surprised to see it as a percentage. Do we see mixed usage like this? I expected at first glance that this would be largely homogeonous: its all JSON or all not. If we can change that assumption perhaps we can parse a few that don't hit the threshold for being JSON, and if so, mark the column as JSON?

robdaemon commented:
> @robdaemon I think we can simply skip fingerprinting BLOBs. Only potentially useful information fingerprint gives is % of nils and I don't think the tradeoff is worth it.

true... but I'm thinking in terms of `TEXT` and `NTEXT` (SQL Server) and such. We can probably skip BLOB since there's no point in fingerprinting binary data, but character data might be different

sbelak commented:
> Can someone give me some context? What actually uses the JSON percentages? And i'm pretty surprised to see it as a percentage. Do we see mixed usage like this? I expected at first glance that this would be largely homogeonous: its all JSON or all not. If we can change that assumption perhaps we can parse a few that don't hit the threshold for being JSON, and if so, mark the column as JSON?

The only place it's use is the corresponding type classifier (see `metabase.sync.analyze.classifiers.text-fingerprint`). We use percentage to be somewhat smart/robust in the face of malformed data. Looking at the code, it could do with a bit of an upgrade in general and eg. ignore nils when counting.

dpsutton commented:
@sbelak good call on checking the other types which map to `:type/Text`. Postgres maps the types `xml', 'jsonb', 'jxon', and some networking types. I've backed the change back to just when `:database_type` is "text". I'm still looking to see if this is something we can rely on that the underlying column will have a substring operation.

dpsutton committed:
Check database_type of text for substring of summaries #13288

in pg, many types go to :type/Text which cannot have substring on
them, such as xml, json, jsonb, etc.

sbelak commented:
Be aware that different DBs name types differently. If you have to wade into drivers, I think it would be better to just define a couple of additional `special_types` and map stuff we don't want to fingerprint to those (ie. `:type/BLOB`), else there's going to be another list of DB-specific names to maintain and we try to steer clear of that and push such functionality/decisions into drivers.

dpsutton committed:
Skip fingerprinting "structured" fields, take substring of text fields

dpsutton committed:
"Sniff" JSON results when truncated in fingerprinting

- had to move the truncation-size var due to circularity, but it most
logically belongs in the fingerprinting namespace as that's the only
mechanism that really should be aware of it anyways

- get rid of :refer :all in a test ns

- fingerprinting doesn't happen on base_type :type/Structured so add
that to the query for fields to fingerprint

dpsutton commented:
should be good to go
- don't fingerprint "structured" fields. this solves if json/xml blobs are large
- substring text fields ( i think this one actually addresses the customer reported problem)
- handle "sniffing" out json fields when truncated but still do parsing when size is lower than `fingerprinters/truncation-size`. This might just be easier to always just sniff json. don't have a strong opinion open to others' thought.

sbelak commented on src/metabase/db/metadata_queries.clj:
```diff
@@ -92,14 +92,24 @@
-    (get-in results [:data :rows])))
+  ([table :- si/TableInstance, fields :- [si/FieldInstance]]
+   (table-rows-sample table fields nil))
+  ([table :- si/TableInstance, fields :- [si/FieldInstance] truncation-size :- (s/maybe s/Int)]
```
I think having an optional options map for things such as `truncation-size` is more future-proof, as I can easily see doing things such as different sampling strategies in the future, which will also need to be signalled

sbelak commented on src/metabase/types.clj:
```diff
@@ -158,6 +158,10 @@
 
+;;; Structured
+
+(derive :type/Structured :type/*)
```
Relevant existing types such as `:type/SerializedJSON` should derive from this.

dpsutton replied:
Serialized JSON is a string? Perhaps it should remain as text and a new `:type/JSON` should exist for json types. Presumably these are actually structured data like xml nodes and json types rather than textual representations of these structures?

sbelak commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -224,11 +225,27 @@
   (u/ignore-exceptions
-    ((some-fn map? sequential?) (json/parse-string x))))
+    (when (and x (string? x))
+      (if (= (.length ^String x) truncation-size)
```
I'd switch to using sniffing only, parsing a ton of JSON is wasteful. Perhaps tweak the filter to first two chars, so we don't pickup literals like `"[foo]"`. Doing so also decouples truncation assumptions from fingerprinters (the way I see it, truncation and even more so the specifics of truncation are a concern of the part of the code that does fingerprinting in the context of sync)

sbelak commented:
Aside from incomplete type hierarchy, this looks good. Can you also open a followup issue to implement the new type for the rest of the drivers.

dpsutton committed:
Move `table-rows-sample` truncation option into options map

dpsutton committed:
Switch json fingerprinting entirely to heuristic rather than parsing

- faster in all cases and accuracy is not a particular goal
- doesn't rely on the truncation size for switching to heuristic

dpsutton committed:
Derive SerializedJson and XML from :type/Structured

dpsutton committed:
Fixes metadata queries test when no truncation-size provided

dpsutton committed:
JSON heuristic for array of booleans

- want to validate as JSON [true, false] but not [random tag]

dpsutton committed:
Removes two unused deps from fingerprinters

we no longer actually parse json so don't need cheshire and don't use
clojure.string/starts-with in favor of a little descent parser

dpsutton commented:
@sbelak i thought i had addressed all requested changes. UI is acting a little weird so i can't see if there are any more pending changes to do. Unless my changes to the type hierarchy are not sufficient/incomplete. I'm creating a new issue to add support for `:type/Structured` and `:type/XML` now.

sbelak commented on src/metabase/db/metadata_queries.clj:
```diff
@@ -89,17 +89,32 @@
   inferring special types and what-not; we don't want to scan millions of values at any rate."
   10000)
 
+(def table-rows-sample-options
```
We use CamelCase for type/schema names

sbelak commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -224,11 +223,27 @@
         :q1  q1
         :q3  q3)))))
 
+(def truncation-size
```
I'd move this to `fingerprint` namespace. There's now nothing in this ns that uses it.

sbelak commented on src/metabase/sync/analyze/fingerprint/fingerprinters.clj:
```diff
@@ -224,11 +223,27 @@
   (u/ignore-exceptions
-    ((some-fn map? sequential?) (json/parse-string x))))
+    (when (and x (string? x))
+      (let [matcher (case (first x)
```
Is this really measurably more performant than a direct/naive regex implementation? I'd just do the first 2 characters without too many assumptions. Look at it like this: it will be faster than what we have, so it's a win. From here I'd optimize for readability & robustness.

dpsutton replied:
i was going to do the first two characters but then we would miss out on `"[true, true]"` is valid JSON but `"[prod]"` is not valid json. that's the only reason. if we want to ignore that case its fine. As I understand it we only care about the first character if its `{` and treat that as a valid map. If its a vector you had mentioned checking its not a single `[tag]` style text.

sbelak replied:
Fair point. Keep in mind though that we require 95% of rows to be valid, so false positives are almost always fine, and we can tolerate at least some false negatives.

sbelak reviewed (approved):
Some minor things, but I don't see them as blocking

dpsutton committed:
CamelCases schema name and move truncation-size to appropriate ns

truncation size used to be used in the fingerprinting analysis when
checking text for JSON (using a heuristic if it was truncated) and
passed into the query to get table summaries. Now its only passed into
the summaries query and the JSON recognizer in the text fingerprinter
always uses heuristics

dpsutton committed:
Fix test now that its not relying on truncation limit for heuristics

previously was making a string that was the `truncation-limit` length
to trigger the heuristics of json parsing. now we always use heuristic
so no need to construct such a precise length string.

rlotun commented:
Let's get this merged in, please @dpsutton - this addresses an issue for our cloud customers cc/ @ascandella who can give you more context, if you're interested.

sbelak commented:
Yes, let's. We just need to make sure #13371 is done before we ship this as otherwise sync will now break on non-Postgres DBs with text-like fields that do not support `substring`.

dpsutton commented:
> Yes, let's. We just need to make sure #13371 is done before we ship this as otherwise sync will now break on non-Postgres DBs with text-like fields that do not support `substring`.

The following drivers talk about `:type/Text`:

```bash
~/p/c/metabase ❯❯❯ ag :type/Text src/metabase/driver
src/metabase/driver/common.clj
255:    String                         :type/Text

src/metabase/driver/sql_jdbc/sync.clj
145:  "Given a `database-type` (e.g. `VARCHAR`) return the mapped Metabase type (e.g. `:type/Text`)."

src/metabase/driver/mysql.clj
236:    :CHAR       :type/Text
246:    :LONGTEXT   :type/Text
249:    :MEDIUMTEXT :type/Text
254:    :TEXT       :type/Text
259:    :TINYTEXT   :type/Text
261:    :VARCHAR    :type/Text

src/metabase/driver/postgres.clj
221:   :bpchar        :type/Text ; "blank-padded char" is the internal name of "character"
225:   :citext        :type/Text ; case-insensitive text
254:   :text          :type/Text
264:   :varchar       :type/Text
267:   (keyword "character varying")          :type/Text

src/metabase/driver/h2.clj
202:   :CHAR                                :type/Text
203:   :CHARACTER                           :type/Text
204:   :CLOB                                :type/Text
222:   :LONGTEXT                            :type/Text
224:   :LONGVARCHAR                         :type/Text
227:   :MEDIUMTEXT                          :type/Text
228:   :NCHAR                               :type/Text
229:   :NCLOB                               :type/Text
230:   :NTEXT                               :type/Text
233:   :NVARCHAR                            :type/Text
234:   :NVARCHAR2                           :type/Text
242:   :TEXT                                :type/Text
247:   :TINYTEXT                            :type/Text
248:   :UUID                                :type/Text
250:   :VARCHAR                             :type/Text
251:   :VARCHAR2                            :type/Text
252:   :VARCHAR_CASESENSITIVE               :type/Text
253:   :VARCHAR_IGNORECASE                  :type/Text

src/metabase/driver/sql/query_processor.clj
501:    (as [:field-literal \"x\" :type/Text])
505:    (as [:datetime-field [:field-literal \"x\" :type/Text] :month])
```

As far as I can tell all types marked `:type/Text` now support substring and we are good to go. That other ticket was really to make use of `:type/Structured` and perhaps another (`:type/BLOB` or `:type/Opaque`) which can't meaningfully be managed. JSON, xml, and the network types can be programmatically manipulated and are good candidates for the structured tag, but we don't use this yet. And the ticket was to start making use of the structured data.

sbelak commented:
We have more drivers in `metabase/modules/drivers`.

dpsutton commented:
looking pretty good there as well

excluding the test namespaces:

```bash
~/p/c/metabase ❯❯❯ ag :type/Text modules/drivers
modules/drivers/presto/src/metabase/driver/presto.clj
58:    #"varchar.*"   :type/Text
59:    #"char.*"      :type/Text
61:    #"json"        :type/Text       ; TODO - this should probably be Dictionary or something

modules/drivers/sparksql/src/metabase/driver/hive_like.clj
47:    #"char.*"           :type/Text
48:    #"varchar.*"        :type/Text
49:    #"string.*"         :type/Text

modules/drivers/oracle/src/metabase/driver/oracle.clj
40:    [#"CHAR"        :type/Text]
41:    [#"CLOB"        :type/Text]
50:    [#"LONG"        :type/Text]
61:    [#"URI"         :type/Text]

modules/drivers/bigquery/src/metabase/driver/bigquery.clj
118:    "STRING"    :type/Text

modules/drivers/druid/src/metabase/driver/druid/sync.clj
16:    "STRING"      :type/Text

modules/drivers/snowflake/src/metabase/driver/snowflake.clj
94:    :VARCHAR                    :type/Text
95:    :CHAR                       :type/Text
96:    :CHARACTER                  :type/Text
97:    :STRING                     :type/Text
98:    :TEXT                       :type/Text

modules/drivers/googleanalytics/src/metabase/driver/googleanalytics/execute.clj
35:                              (= data-type "STRING")    :type/Text)]
48:  {"STRING"      :type/Text

modules/drivers/sqlite/src/metabase/driver/sqlite.clj
62:    [#"CHAR"      :type/Text]
63:    [#"TEXT"      :type/Text]
64:    [#"CLOB"      :type/Text]

modules/drivers/vertica/src/metabase/driver/vertica.clj
35:     :Char                      :type/Text
36:     :Varchar                   :type/Text
47:     (keyword "Long Varchar")   :type/Text

modules/drivers/sqlserver/src/metabase/driver/sqlserver.clj
39:    :char             :type/Text
53:    :nchar            :type/Text
54:    :ntext            :type/Text
56:    :nvarchar         :type/Text
63:    :text             :type/Text
69:    :varchar          :type/Text
```

looks like perhaps we could change presto's json column to `:type/SerialzedJSON` or structured depending on if the driver returns a map or a string.

sbelak commented:
Awesome. This should be pretty low-drama to add then. Untangling presto is a stretch goal, but the focus is on getting this fix shippable as it will make a huge difference for a lot of our users.

dpsutton commented:
>  but the focus is on getting this fix shippable as it will make a huge difference for a lot of our users.

I'm not aware of any blockers then. My understanding is that we are good to merge
